clarity/contracts/vaults-v2/arkadiko-vaults-sorted-v1-1.clar

```diff
@@ -0,0 +1,279 @@
+;; Vaults Sorted 
+;; Keep vaults sorted based in NICR (nominal collateral ratio) 
+;;
+
+;; ---------------------------------------------------------
+;; Constants
+;; ---------------------------------------------------------
+
+(define-constant ERR_NOT_AUTHORIZED u960401)
+(define-constant ERR_WRONG_POSITION u960001)
+
+;; ---------------------------------------------------------
+;; Maps
+;; ---------------------------------------------------------
+
+(define-map tokens 
+  { 
+    token: principal  
+  } 
+  {
+    first-owner: (optional principal),
+    last-owner: (optional principal),
+    total-vaults: uint
+  }
+)
+
+(define-map vaults 
+  { 
+    owner: principal,
+    token: principal  
+  } 
+  {
+    prev-owner: (optional principal),
+    next-owner: (optional principal),
+    nicr: uint
+  }
+)
+
+;; ---------------------------------------------------------
+;; Getters
+;; ---------------------------------------------------------
+
+(define-read-only (get-token (token principal))
+  (default-to
+    {
+      first-owner: none,
+      last-owner: none,
+      total-vaults: u0
+    }
+    (map-get? tokens { token: token })
+  )
+)
+
+(define-read-only (get-vault (owner principal) (token principal))
+  (map-get? vaults { owner: owner, token: token })
+)
+
+(define-read-only (has-access (caller principal))
+  (or
+    (is-eq caller (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "vaults-operations")))
+    (is-eq caller (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "vaults-manager")))
+    (is-eq caller (contract-call? .arkadiko-dao get-dao-owner))
+  )
+)
+
+;; ---------------------------------------------------------
+;; Find position
+;; ---------------------------------------------------------
+
+;; TODO: check the cost
```

Comment:
Q: does it make sense to figure this out first? 

What's the alternative if the costs are too high?

Reply:
A: Costs in unit tests look good now. Have not thought about an alternative yet.